### PR TITLE
rework interface pruning

### PR DESCRIPTION
**Pull request by v4hn**
@rhaschke Here's the big patch I talked about yesterday.

I added/adjusted quite a few tests, with minor improvements on the side.

Sorry it all got mashed up in one huge commit but I wanted to omit functionally broken commits and most of this patch is interrelated.

a few minor changes are left as standalone commits before the main one.

### CODE REVIEW

**Comment by v4hn** [comment]
Is there a way to have Codecov report metrics relative to the *total* number of lines in the resulting project? I always feel the reports are unfair to patches that decrease the total number of code lines.

**Comment by rhaschke** [comment]
I thoroughly reviewed and revised the interface resolution. Thanks for initiating this rework.

**Comment by rhaschke** [comment]
Fixes #131.

File: core/src/container.cpp [hunk 1]
```diff
@@ -425,51 +425,39 @@ void SerialContainerPrivate::connect(StagePrivate& stage1, StagePrivate& stage2)
...
+	// connect last child's (end) push interface
 	setChildsPushForwardInterface(last);
-
-	if ((accepted & END_IF_MASK) != UNKNOWN &&
-	    (last.pimpl()->requiredInterface() & END_IF_MASK) != (accepted & END_IF_MASK)) {
-		boost::format desc(
-		    "requested end interface for container (%1%) does not agree with inferred end interface of last child (%2%)");
-		desc % flowSymbol<END_IF_MASK>(accepted) % flowSymbol<END_IF_MASK>(last.pimpl()->requiredInterface());
-		throw InitStageException(*me(), desc.str());
-	}
```
Comment (v4hn) [inline core/src/container.cpp:606]: Is the end interface still guaranteed to be the requested one?

File: core/src/stage.cpp [hunk 2]
```diff
@@ -112,28 +112,6 @@ InterfaceFlags StagePrivate::interfaceFlags() const {
...
-	// only one side needs to be specified but the value has to be compatible with this stage
-	if (((accepted_start != UNKNOWN) && (accepted_start & required_start) != required_start) ||
-	    ((accepted_end != UNKNOWN) && (accepted_end & required_end) != required_end)) {
-		boost::format desc("interface %1% %2% does not match inferred interface %3% %4%");
-		desc % flowSymbol<START_IF_MASK>(required_start) % flowSymbol<END_IF_MASK>(required_end);
-		desc % flowSymbol<START_IF_MASK>(accepted_start) % flowSymbol<END_IF_MASK>(accepted_end);
-		throw InitStageException(*me(), desc.str());
-	}
-}
-
```
Comment (v4hn) [inline core/src/stage.cpp]: Why did you remove the check?
I don't see harm in testing all stages during setup time, although of course the invariant *should* hold for simple interfaces.

File: demo/scripts/grasp.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+#! /usr/bin/env python
```
Comment (v4hn) [inline demo/scripts/grasp.py:1]: You added `demo/scripts/grasp.py`, `demo/scripts/pick.py`, `notes.org`, `pick+place.org` and `test.py` to the repo in this commit.

I guess the last three were committed by accident?

**Comment by v4hn** [comment]
I just looked through your changes. It sure makes sense to unify ~pruning~resolving and validation.
Hope you had a Good Friday (capital letters intended). Thanks for spending your spare time on this.

File: demo/scripts/grasp.py (hunk 3 above)
Comment (rhaschke) [reply demo/scripts/grasp.py:1]: Oops. That was an accident indeed. If you don't mind, I would like to rewrite history :wink:

Comment (v4hn) [reply demo/scripts/grasp.py:1]: You know my opinion on force pushes.
This has been upstream for 4 days and we apparently have an active user base these days.
If you consider your org-mode files private information, feel free to go ahead.
Otherwise I would simply add a removal-commit.
